libcaf_net/caf/net/stream_transport.hpp

```diff
@@ -198,49 +207,101 @@ class stream_transport {
     prepare_next_read();
   }
 
-  void write_packet(span<const byte> header, span<const byte> payload,
-                    typename worker_type::id_type) {
-    if (write_buf_.empty())
+  void write_packet(unit_t, span<buffer_type*> buffers) {
+    // Sanity check
+    CAF_ASSERT(!buffers.empty());
+    auto it = buffers.begin();
+    if (write_queue_.empty())
       manager().register_writing();
-    write_buf_.insert(write_buf_.end(), header.begin(), header.end());
-    write_buf_.insert(write_buf_.end(), payload.begin(), payload.end());
+    // move header by itself to keep things sorted.
+    write_queue_.emplace_back(true, std::move(**it++));
+    // payload buffers. just write them
+    for (; it != buffers.end(); ++it)
+      write_queue_.emplace_back(false, std::move(**it));
+  }
+
+  // -- buffer recycling -------------------------------------------------------
+
+  buffer_type next_header_buffer() {
+    return next_buffer_impl(free_header_bufs_);
+  }
+
+  buffer_type next_buffer() {
+    return next_buffer_impl(free_bufs_);
+  }
+
+  buffer_type next_buffer_impl(std::deque<buffer_type>& container) {
+    if (container.empty()) {
+      return {};
+    } else {
+      auto buf = std::move(container.front());
+      container.pop_front();
+      return buf;
+    }
   }
 
 private:
   // -- private member functions -----------------------------------------------
 
   bool write_some() {
-    if (write_buf_.empty())
+    CAF_LOG_TRACE(CAF_ARG(handle_.id));
+    // helper to sort empty buffers back into the right queues
+    auto recycle = [&]() {
+      auto& front = write_queue_.front();
+      auto& is_header = front.first;
+      auto& buf = front.second;
+      written_ = 0;
+      buf.clear();
+      if (is_header && free_header_bufs_.size() < max_header_bufs_)
+        free_header_bufs_.emplace_back(std::move(buf));
+      else if (free_bufs_.size() < max_output_bufs_)
+        free_bufs_.emplace_back(std::move(buf));
+      write_queue_.pop_front();
+    };
+    // nothing to write
+    if (write_queue_.empty())
       return false;
-    auto len = write_buf_.size() - written_;
-    auto buf = write_buf_.data() + written_;
-    CAF_LOG_TRACE(CAF_ARG(handle_.id) << CAF_ARG(len));
-    auto ret = write(handle_, make_span(buf, len));
-    if (auto num_bytes = get_if<size_t>(&ret)) {
-      CAF_LOG_DEBUG(CAF_ARG(len) << CAF_ARG(handle_.id) << CAF_ARG(*num_bytes));
-      // Update state.
-      written_ += *num_bytes;
-      if (written_ >= write_buf_.size()) {
-        written_ = 0;
-        write_buf_.clear();
-        return false;
+    do {
+      auto& buf = write_queue_.front().second;
+      if (buf.empty()) {
```

Comment:
Can we drop this branch and instead use `CAF_ASSERT(!buf.empty())`?

Reply:
I don't think so. At least not with the current test-implementations. Some tests still send empty buffers I think.

Reply:
Should I add that change to this PR or can I move that to another one?

Reply:
Done.